lib/services/access_list.go

```diff
@@ -243,20 +257,141 @@ func IsAccessListOwner(identity tlsca.Identity, accessList *accesslist.AccessLis
 // AccessListMembershipChecker will check if users are members of an access list and
 // makes sure the user is not locked and meets membership requirements.
 type AccessListMembershipChecker struct {
-	members AccessListMemberGetter
+	members AccessListsAndMembersGetter
 	locks   LockGetter
 	clock   clockwork.Clock
 }
 
 // NewAccessListMembershipChecker will create a new access list membership checker.
-func NewAccessListMembershipChecker(clock clockwork.Clock, members AccessListMemberGetter, locks LockGetter) *AccessListMembershipChecker {
+func NewAccessListMembershipChecker(clock clockwork.Clock, members AccessListsAndMembersGetter, locks LockGetter) *AccessListMembershipChecker {
 	return &AccessListMembershipChecker{
 		members: members,
 		locks:   locks,
 		clock:   clock,
 	}
 }
 
+func getAccessListDynamicMembers(ctx context.Context, clt AccessListsAndMembersGetter, entry string) ([]string, error) {
+	var pageToken string
+	var dynamicMembers []string
+
+	for {
+		var members []*accesslist.AccessListMember
+		var err error
+
+		members, pageToken, err = clt.ListAccessListMembers(ctx, entry, 0 /* default page size */, pageToken)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		for _, member := range members {
+			if member.Spec.MembershipKind == accesslistv1.MembershipKind_MEMBERSHIP_KIND_LIST.String() {
+				dynamicMembers = append(dynamicMembers, member.GetName())
+			}
+		}
+
+		if pageToken == "" {
+			break
+		}
+	}
+
+	return dynamicMembers, nil
+}
+
+func (a AccessListMembershipChecker) recursiveIsAccessListMemberCheck(ctx context.Context, identity tlsca.Identity, accessList *accesslist.AccessList) error {
+	seen := map[string]struct{}{}
+	queue := []string{accessList.GetName()}
+
+	for len(queue) > 0 {
+		pal := queue[0]
+		queue = queue[1:]
+
+		member, err := a.members.GetAccessListMember(ctx, pal, identity.Username)
+		if err != nil && !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+		if trace.IsNotFound(err) {
+			subAccessListMembers, err := getAccessListDynamicMembers(ctx, a.members, pal)
+			if err != nil {
+				return trace.NotFound("error finding access list %s", pal)
+			}
+			for _, next := range subAccessListMembers {
+				if _, ok := seen[next]; ok {
+					continue
+				}
+				seen[next] = struct{}{}
+				queue = append(queue, next)
+			}
+			continue
+		}
+
+		expires := member.Spec.Expires
+		if !expires.IsZero() && !a.clock.Now().Before(expires) {
+			return trace.AccessDenied("user %s's membership has expired in the access list", identity.Username)
```

Comment:
I think there's a bug here, although, I think we'd want to check with someone what the expected behaviour should be:
- UserA is a member of AccessListA and AccessListB
- AccessListA and AccessListB are members of AccessListC
- UserA's membership of AccessListA is expired, but their membership of AccessListB is not.

It looks to me that when calculating membership of AccessListC:
- If AccessListA is processed first, then they receive this error.
- If AccessListB is processed first, then they are granted access

It probably ought to be deterministic and that will mean that even if we encounter an expired membership, we need to continue searching the tree.